qualtran/bloqs/bookkeeping/join.py

```diff
@@ -81,27 +81,18 @@ def adjoint(self) -> 'Bloq':
     def as_cirq_op(self, qubit_manager, reg: 'CirqQuregT') -> Tuple[None, Dict[str, 'CirqQuregT']]:
         return None, {'reg': reg.reshape(self.dtype.num_qubits)}
 
-    def add_my_tensors(
-        self,
-        tn: 'qtn.TensorNetwork',
-        tag: Any,
-        *,
-        incoming: Dict[str, 'SoquetT'],
-        outgoing: Dict[str, 'SoquetT'],
-    ):
+    def my_tensors(
+        self, incoming: Dict[str, 'ConnectionT'], outgoing: Dict[str, 'ConnectionT']
+    ) -> List['qtn.Tensor']:
         import quimb.tensor as qtn
 
-        if not isinstance(incoming['reg'], np.ndarray):
-            raise ValueError('Incoming register must be a numpy array')
-        tn.add(
-            qtn.Tensor(
-                data=np.eye(2**self.dtype.num_qubits, 2**self.dtype.num_qubits).reshape(
-                    (2,) * self.dtype.num_qubits + (2**self.dtype.num_qubits,)
-                ),
-                inds=incoming['reg'].tolist() + [outgoing['reg']],
-                tags=['Join', tag],
-            )
-        )
+        eye = np.eye(2)
+        incoming = cast('NDArray', incoming['reg'])
+        outgoing = outgoing['reg']
+        return [
+            qtn.Tensor(data=eye, inds=[(outgoing, i), (incoming[i], 0)], tags=[str(self)])
+            for i in range(self.dtype.num_qubits)
+        ]
```

Comment:
When implementing the `my_tensors` method, a pattern like this looks super confusing. The indices now have two types of indexing going on - indexing into a numpy array if `self.reg.shape` to get the corresponding `Connection` object and then an implicit indexing into an implicit array of size `self.bitsize` by specifying a tuple `(cxn, i)`

I think this calls for adding a new type (equivalent to a qubit) so that the `ConnectionT` for a register `reg` would now be a numpy array of shape `reg.shape + (reg.bitsize,)`. This has the following advantages:
- It gets rid of an implicit vs explicit indexing. The implicit indexing is confusing at best.
- The concept aligns with the concept of a qubit - we just say that each entry in the incoming and outgoing dictionary corresponds to a qubit. 
-  This kind of `full shape` is already used as a concept in cirq style API of `decompose_from_registers` and would so it'll unify the tensor contraction API nicely with that API. 


Another alternative would be to keep `ConnectionT` a numpy array of shape `reg.shape` but override the `__getitem__` method on `Connection` object so `cxn[i]` returns a new object which is essentially a dataclass / tuple storing (cxn, i) and can be used as a tensor index (like `TensorIndex(cxn, i)`)

In general, it would be much nicer if the call sites look like `inds=[outgoing[i], incoming[i][0]]` instead of what it is right now. 

What do you think?

Reply:
There is three levels of indexing: by register name, optionally within a "shaped" register, and then the sub-register bit-indexing. 

I don't think it's worth adding an extra dimension to all the arrays where the last column is just the numbers 0..n. 

`(incoming['ctrl'], 0)` is not meaningfully more confusing than `incoming['ctrl'][0]` where indeed the latter syntax just maps to `TensorIndex(incoming['ctrl'], 0)` anyways